Content.Shared/_RMC14/Xenonids/Leap/XenoLeapSystem.cs

```diff
@@ -217,6 +228,157 @@ private void OnXenoLeapingPullAttempt(Entity<XenoLeapingComponent> ent, ref Pull
+    {
+        if (ent.Comp.InherentStunDuration == null)
+            return;
+
+        ent.Comp.StunDuration = ent.Comp.InherentStunDuration.Value;
+    }
+
+    /// <summary>
+    ///     Apply the <see cref="RMCLeapProtectionComponent"/> to the given entity.
+    /// </summary>
+    /// <param name="receiver">The entity that receives the <see cref="RMCLeapProtectionComponent"/></param>
+    /// <param name="protection">The entity that provides the component using <see cref="RMCGrantLeapProtectionComponent"/></param>
+    private void ApplyLeapProtection(EntityUid receiver, Entity<RMCGrantLeapProtectionComponent> protection)
+    {
+        var leapProtection = EnsureComp<RMCLeapProtectionComponent>(receiver);
+        leapProtection.ProtectionProviders.Add(protection);
+
+        if (protection.Comp.StunDuration >= leapProtection.StunDuration)
+        {
+            leapProtection.StunDuration = protection.Comp.StunDuration;
+            leapProtection.BlockSound = protection.Comp.BlockSound;
+        }
+
+        Dirty(receiver, leapProtection);
+    }
+
+    /// <summary>
+    ///     Remove the entity with the <see cref="RMCGrantLeapProtectionComponent"/> from the list of entities that
+    ///     provide the <see cref="RMCLeapProtectionComponent"/> to the user.
+    ///     Then check if there are any other entities equipped that provide the component.
+    /// </summary>
+    /// <param name="user">The entity that unequipped an entity with <see cref="RMCGrantLeapProtectionComponent"/></param>
+    /// <param name="protection">The entity that got unequipped</param>
+    /// <returns>If the entity should keep the <see cref="RMCLeapProtectionComponent"/></returns>
+    private bool RemoveLeapProtection(EntityUid user, Entity<RMCGrantLeapProtectionComponent> protection)
+    {
+        if (!TryComp(user, out RMCLeapProtectionComponent? leapProtection))
+            return true;
+
+        var stunDuration = new TimeSpan();
+        leapProtection.ProtectionProviders.Remove(protection);
+
+        // Set the inherent stun duration if it exists.
+        if (leapProtection.InherentStunDuration != null)
+        {
+            stunDuration = leapProtection.InherentStunDuration.Value;
+            leapProtection.InherentBlockSound = leapProtection.InherentBlockSound;
+        }
+
+        // Check if there are any other entities equipped that provide leap protection.
+        foreach (var protectionGranter in leapProtection.ProtectionProviders)
+        {
+            if (!TryComp(protectionGranter, out RMCGrantLeapProtectionComponent? grantProtection) ||
+                grantProtection.StunDuration < stunDuration)
+                continue;
+
+            stunDuration = grantProtection.StunDuration;
+            leapProtection.BlockSound = grantProtection.BlockSound;
+        }
+
+        // Don't remove the component if it's inherent or if another component provides protection.
+        if (stunDuration != new TimeSpan())
```

Comment:
This is clearer for comparing against a time of zero
```suggestion
        if (stunDuration != TimeSpan.Zero)
```